k2/python/tests/random_paths_test.py

```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+#
+# Copyright (c)  2021  Xiaomi Corp.       (authors: Fangjun Kuang)
+#
+# See ../../../LICENSE for clarification regarding multiple authors
+
+# To run this single test, use
+#
+#  ctest --verbose -R random_paths_test_py
+
+import unittest
+
+import k2.sparse
+import torch
+
+
+class TestRandomPaths(unittest.TestCase):
+
+    def test_single_fsa_case1(self):
+        devices = [torch.device('cpu')]
+        if torch.cuda.is_available():
+            devices.append(torch.device('cuda', 0))
+
+        for device in devices:
+            for use_double_scores in (True, False):
+                s = '''
+                    0 1 1 0.1
+                    1 2 -1 0.2
+                    2
+                '''
+                fsa = k2.Fsa.from_str(s).to(device)
+                fsa_vec = k2.create_fsa_vec([fsa])
+                path = k2.random_paths(fsa_vec,
+                                       use_double_scores=use_double_scores,
+                                       num_paths=2)
+                assert path.num_axes() == 3
+                self.assertEqual(str(path), '[ [ [ 0 1 ] [ 0 1 ] ] ]')
+
+    def test_single_fsa_case2(self):
+        devices = [torch.device('cpu')]
+        if torch.cuda.is_available():
+            devices.append(torch.device('cuda', 0))
+
+        for device in devices:
+            for use_double_scores in (True, False):
+                s = '''
+                    0 1 1 1
+                    0 1 2 1.1
+                    1 2 3 1
+                    1 2 4 1.1
+                    2 3 -1 1
+                    2 3 -1 1.1
+                    3
+                '''
+                fsa = k2.Fsa.from_str(s).to(device)
+                fsa_vec = k2.create_fsa_vec([fsa])
+
+                path = k2.random_paths(fsa_vec,
+                                       use_double_scores=use_double_scores,
+                                       num_paths=1)
+                assert path.num_axes() == 3
+                # p is 0.5, it should select the last leaving arc of each state
+                self.assertEqual(str(path), '[ [ [ 1 3 5 ] ] ]')
```

Comment:
@danpovey 
CPU gives [ 1, 3, 5], while CUDA returns [1, 2, 4].

Reply:
Your test logic is not quite right-- it "zooms in" to the distribution each time it takes an arc-- but I did find a bug.  Thanks!  Will push a fix.

Reply:
.. the CPU code was failing to zoom in.

Reply:
> Your test logic is not quite right

My test code is written according to the CPU version. Will update it once you fix it.